File: apps/web/test/fixtures/plants.ts
import type {
  PagePlantResponse,
  PagePlantTimelineEntryResponse,
  PlantImageResponse,
  PlantResponse,
  PlantTimelineEntryResponse,
  PlantTransferResponse,
} from "@/lib/api/plants";
import type {
  EnvironmentalRecordResponse,
  FertilizerRecordResponse,
  GrowthRecordResponse,
  HealthRecordResponse,
  PageEnvironmentalRecordResponse,
  PageFertilizerRecordResponse,
  PageGrowthRecordResponse,
  PageHealthRecordResponse,
  PageWateringRecordResponse,
  WateringRecordResponse,
} from "@/lib/api/plant-records";
import type { DiseaseReportResponse, TreatmentResponse } from "@/lib/api/disease-reports";

/** Shared fixtures for 7G Plant Lifecycle tests -- mirrors test/fixtures/catalog.ts's pattern. */

export function makePlant(overrides: Partial<PlantResponse> = {}): PlantResponse {
  return {
    id: "cccccccc-cccc-cccc-cccc-cccccccccc01",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: "44444444-4444-4444-4444-444444444444",
    species_id: "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaa01",
    variety_id: null,
    common_label: "Bench 3 - Fig #1",
    zone: "Greenhouse A",
    status: "in_production",
    qr_code_token: "qr-token-plant-1",
    price: 24.99,
    planted_at: "2026-01-01T00:00:00Z",
    sold_at: null,
    deceased_at: null,
    deceased_reason: null,
    batch_number: "B-2026-001",
    supplier_id: null,
    purchase_price: null,
    purchase_date: null,
    registered_by_user_id: "11111111-1111-1111-1111-111111111111",
    archived_at: null,
    archived_reason: null,
    created_at: "2026-01-01T00:00:00Z",
    updated_at: "2026-01-01T00:00:00Z",
    age_days: 30,
    ...overrides,
  } as PlantResponse;
}

export function makePlantPage(items: PlantResponse[] = [makePlant()]): PagePlantResponse {
  return {
    items,
    meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 },
  };
}

export function makeGrowthRecord(overrides: Partial<GrowthRecordResponse> = {}): GrowthRecordResponse {
  return {
    id: "dddddddd-dddd-dddd-dddd-dddddddddd01",
    plant_id: makePlant().id,
    height_cm: 45,
    spread_cm: 30,
    leaf_count: 12,
    flower_count: 0,
    fruit_count: 0,
    growth_stage: "growing",
    photo_url: null,
    photo_urls: null,
    notes: "Healthy new growth.",
    recorded_by_user_id: "11111111-1111-1111-1111-111111111111",
    recorded_at: "2026-08-01T00:00:00Z",
    ...overrides,
  } as GrowthRecordResponse;
}

export function makeGrowthPage(items: GrowthRecordResponse[] = [makeGrowthRecord()]): PageGrowthRecordResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeHealthRecord(overrides: Partial<HealthRecordResponse> = {}): HealthRecordResponse {
  return {
    id: "eeeeeeee-eeee-eeee-eeee-eeeeeeeeee01",
    plant_id: makePlant().id,
    status_label: "healthy",
    health_score: 92,
    notes: null,
    photo_url: null,
    is_ai_observation: false,
    recorded_by_user_id: "11111111-1111-1111-1111-111111111111",
    recorded_at: "2026-08-01T00:00:00Z",
    ...overrides,
  } as HealthRecordResponse;
}

export function makeHealthPage(items: HealthRecordResponse[] = [makeHealthRecord()]): PageHealthRecordResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeWateringRecord(overrides: Partial<WateringRecordResponse> = {}): WateringRecordResponse {
  return {
    id: "ffffffff-ffff-ffff-ffff-ffffffffff01",
    plant_id: makePlant().id,
    branch_id: "44444444-4444-4444-4444-444444444444",
    zone: "Greenhouse A",
    volume_ml: 250,
    method: "drip",
    notes: null,
    recorded_by_user_id: "11111111-1111-1111-1111-111111111111",
    recorded_at: "2026-08-01T00:00:00Z",
    ...overrides,
  } as WateringRecordResponse;
}

export function makeWateringPage(items: WateringRecordResponse[] = [makeWateringRecord()]): PageWateringRecordResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeFertilizerRecord(overrides: Partial<FertilizerRecordResponse> = {}): FertilizerRecordResponse {
  return {
    id: "11111111-2222-3333-4444-555555555501",
    plant_id: makePlant().id,
    branch_id: "44444444-4444-4444-4444-444444444444",
    zone: "Greenhouse A",
    product_name: "Balanced 10-10-10",
    quantity_ml: 50,
    npk_ratio: "10-10-10",
    method: "soil drench",
    schedule: null,
    next_application_date: null,
    notes: null,
    recorded_by_user_id: "11111111-1111-1111-1111-111111111111",
    recorded_at: "2026-08-01T00:00:00Z",
    ...overrides,
  } as FertilizerRecordResponse;
}

export function makeFertilizerPage(items: FertilizerRecordResponse[] = [makeFertilizerRecord()]): PageFertilizerRecordResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeEnvironmentalRecord(overrides: Partial<EnvironmentalRecordResponse> = {}): EnvironmentalRecordResponse {
  return {
    id: "22222222-3333-4444-5555-666666666601",
    plant_id: makePlant().id,
    branch_id: "44444444-4444-4444-4444-444444444444",
    zone: "Greenhouse A",
    temperature_celsius: 22,
    humidity_percent: 55,
    soil_moisture_percent: 40,
    light_lux: 8000,
    ph_level: 6.5,
    weather_snapshot: null,
    source: "manual",
    recorded_at: "2026-08-01T00:00:00Z",
    ...overrides,
  } as EnvironmentalRecordResponse;
}

export function makeEnvironmentalPage(
  items: EnvironmentalRecordResponse[] = [makeEnvironmentalRecord()],
): PageEnvironmentalRecordResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeDiseaseReport(overrides: Partial<DiseaseReportResponse> = {}): DiseaseReportResponse {
  return {
    id: "33333333-4444-5555-6666-777777777701",
    plant_id: makePlant().id,
    condition_name: "Root rot",
    status: "draft",
    severity: "medium",
    is_ai_sourced: false,
    ai_confidence: null,
    photo_url: null,
    confirmed_by_user_id: null,
    confirmed_at: null,
    dismissed_reason: null,
    created_at: "2026-08-01T00:00:00Z",
    ...overrides,
  } as DiseaseReportResponse;
}

export function makeTreatment(overrides: Partial<TreatmentResponse> = {}): TreatmentResponse {
  return {
    id: "44444444-5555-6666-7777-888888888801",
    disease_report_id: makeDiseaseReport().id,
    description: "Reduced watering frequency and repotted with fresh soil.",
    outcome: "ongoing",
    applied_by_user_id: "11111111-1111-1111-1111-111111111111",
    applied_at: "2026-08-02T00:00:00Z",
    ...overrides,
  };
}

export function makePlantImage(overrides: Partial<PlantImageResponse> = {}): PlantImageResponse {
  return {
    id: "55555555-6666-7777-8888-999999999901",
    plant_id: makePlant().id,
    url: "https://example.com/photo.jpg",
    thumbnail_url: null,
    caption: "First photo",
    captured_at: "2026-08-01T00:00:00Z",
    uploaded_by_user_id: "11111111-1111-1111-1111-111111111111",
    ...overrides,
  };
}

export function makePlantTransfer(overrides: Partial<PlantTransferResponse> = {}): PlantTransferResponse {
  return {
    id: "66666666-7777-8888-9999-aaaaaaaaaa01",
    plant_id: makePlant().id,
    from_branch_id: "44444444-4444-4444-4444-444444444444",
    to_branch_id: "77777777-7777-7777-7777-777777777777",
    from_zone: "Greenhouse A",
    to_zone: "Greenhouse B",
    note: "Rebalancing stock.",
    transferred_by_user_id: "11111111-1111-1111-1111-111111111111",
    transferred_at: "2026-08-05T00:00:00Z",
    ...overrides,
  };
}

export function makeTimelineEntry(overrides: Partial<PlantTimelineEntryResponse> = {}): PlantTimelineEntryResponse {
  return {
    event_type: "plant.registered",
    occurred_at: "2026-01-01T00:00:00Z",
    summary: "Plant registered at Main Branch.",
    source_id: makePlant().id,
    actor_user_id: "11111111-1111-1111-1111-111111111111",
    ...overrides,
  };
}

export function makeTimelinePage(items: PlantTimelineEntryResponse[] = [makeTimelineEntry()]): PagePlantTimelineEntryResponse {
  return { items, meta: { page: 1, page_size: 30, total_items: items.length, total_pages: 1 } };
}

File: apps/api/app/schemas/plants.py
"""Pydantic request/response DTOs for Module 6 (Plant Lifecycle Management)."""
from __future__ import annotations

import uuid
from datetime import datetime, timezone

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.db.enums import DiseaseReportSeverity, DiseaseReportStatus, PlantStatus, TreatmentOutcome

# ==============================================================================
# Plant Registration / Profile
# ==============================================================================


class RegisterPlantRequest(BaseModel):
    branch_id: uuid.UUID
    species_id: uuid.UUID
    variety_id: uuid.UUID | None = None
    common_label: str | None = Field(None, max_length=255)
    zone: str | None = Field(None, max_length=100)
    batch_number: str | None = Field(None, max_length=100)
    supplier_id: uuid.UUID | None = None
    purchase_price: float | None = Field(None, ge=0)
    purchase_date: datetime | None = None
    price: float | None = Field(None, ge=0)
    planted_at: datetime | None = None
    description: str | None = Field(None, max_length=5000)


class BulkRegisterPlantsRequest(BaseModel):
    plants: list[RegisterPlantRequest] = Field(..., min_length=1, max_length=500)


class UpdatePlantProfileRequest(BaseModel):
    common_label: str | None = Field(None, max_length=255)
    variety_id: uuid.UUID | None = None
    batch_number: str | None = Field(None, max_length=100)
    supplier_id: uuid.UUID | None = None
    purchase_price: float | None = Field(None, ge=0)
    purchase_date: datetime | None = None
    price: float | None = Field(None, ge=0)
    description: str | None = Field(None, max_length=5000)


class TransitionStatusRequest(BaseModel):
    to_status: PlantStatus
    reason: str | None = None


class MovePlantRequest(BaseModel):
    to_branch_id: uuid.UUID | None = None
    to_zone: str | None = Field(None, max_length=100)
    note: str | None = None


class ArchivePlantRequest(BaseModel):
    reason: str | None = None


class UploadPlantImageRequest(BaseModel):
    url: str = Field(..., min_length=1, max_length=1000)
    thumbnail_url: str | None = Field(None, max_length=1000)
    caption: str | None = Field(None, max_length=255)


class PlantImageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    url: str
    thumbnail_url: str | None
    caption: str | None
    captured_at: datetime
    uploaded_by_user_id: uuid.UUID | None


class PlantResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    species_id: uuid.UUID
    variety_id: uuid.UUID | None
    common_label: str | None
    zone: str | None
    status: PlantStatus
    qr_code_token: str
    price: float | None
    planted_at: datetime
    sold_at: datetime | None
    deceased_at: datetime | None
    deceased_reason: str | None
    batch_number: str | None
    supplier_id: uuid.UUID | None
    purchase_price: float | None
    purchase_date: datetime | None
    registered_by_user_id: uuid.UUID | None
    archived_at: datetime | None
    archived_reason: str | None
    description: str | None
    created_at: datetime
    updated_at: datetime

    @computed_field  # type: ignore[prop-decorator]
    @property
    def age_days(self) -> int:
        """Plant Profile's "Age" -- always derived from `planted_at`, never stored (there is nothing to keep in sync), and serialized into every PlantResponse via Pydantic v2's `@computed_field`."""
        planted = self.planted_at if self.planted_at.tzinfo else self.planted_at.replace(tzinfo=timezone.utc)
        return (datetime.now(timezone.utc) - planted).days


class PlantTransferResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    from_branch_id: uuid.UUID
    to_branch_id: uuid.UUID
    from_zone: str | None
    to_zone: str | None
    note: str | None
    transferred_by_user_id: uuid.UUID
    transferred_at: datetime


# ==============================================================================
# Growth / Health / Watering / Fertilizer / Environmental Records
# ==============================================================================


class RecordGrowthRequest(BaseModel):
    height_cm: float | None = Field(None, ge=0)
    spread_cm: float | None = Field(None, ge=0)
    leaf_count: int | None = Field(None, ge=0)
    flower_count: int | None = Field(None, ge=0)
    fruit_count: int | None = Field(None, ge=0)
    growth_stage: str | None = Field(None, max_length=50, description="Free-text, e.g. 'seedling', 'growing', 'mature'")
    notes: str | None = None
    photo_urls: list[str] | None = None
    measured_at: datetime | None = None


class GrowthRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    height_cm: float | None
    spread_cm: float | None
    leaf_count: int | None
    flower_count: int | None
    fruit_count: int | None
    growth_stage: str | None
    photo_url: str | None
    photo_urls: list | None
    notes: str | None
    recorded_by_user_id: uuid.UUID
    recorded_at: datetime


class RecordHealthRequest(BaseModel):
    status_label: str = Field(..., min_length=1, max_length=50)
    health_score: float | None = Field(None, ge=0, le=100)
    notes: str | None = None
    photo_url: str | None = Field(None, max_length=1000)
    is_ai_observation: bool = False
    observed_at: datetime | None = None


class HealthRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    status_label: str
    health_score: float | None
    notes: str | None
    photo_url: str | None
    is_ai_observation: bool
    recorded_by_user_id: uuid.UUID
    recorded_at: datetime


class RecordWateringRequest(BaseModel):
    volume_ml: float | None = Field(None, ge=0)
    method: str | None = Field(None, max_length=50)
    notes: str | None = None
    watered_at: datetime | None = None


class WateringRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    branch_id: uuid.UUID
    zone: str | None
    volume_ml: float | None
    method: str | None
    notes: str | None
    recorded_by_user_id: uuid.UUID
    recorded_at: datetime


class RecordFertilizerRequest(BaseModel):
    product_name: str = Field(..., min_length=1, max_length=255)
    quantity_ml: float | None = Field(None, ge=0)
    npk_ratio: str | None = Field(None, max_length=20)
    method: str | None = Field(None, max_length=50)
    schedule: str | None = Field(None, max_length=50)
    next_application_date: datetime | None = None
    notes: str | None = None
    applied_at: datetime | None = None


class FertilizerRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    branch_id: uuid.UUID
    zone: str | None
    product_name: str
    quantity_ml: float | None
    npk_ratio: str | None
    method: str | None
    schedule: str | None
    next_application_date: datetime | None
    notes: str | None
    recorded_by_user_id: uuid.UUID
    recorded_at: datetime


class RecordEnvironmentalRequest(BaseModel):
    temperature_celsius: float | None = None
    humidity_percent: float | None = Field(None, ge=0, le=100)
    soil_moisture_percent: float | None = Field(None, ge=0, le=100)
    light_lux: float | None = Field(None, ge=0)
    ph_level: float | None = Field(None, ge=0, le=14)
    weather_snapshot: dict | None = None
    source: str = Field("manual", max_length=20)
    recorded_at: datetime | None = None


class EnvironmentalRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    branch_id: uuid.UUID
    zone: str | None
    temperature_celsius: float | None
    humidity_percent: float | None
    soil_moisture_percent: float | None
    light_lux: float | None
    ph_level: float | None
    weather_snapshot: dict | None
    source: str
    recorded_at: datetime


# ==============================================================================
# Disease Reports / Treatments
# ==============================================================================


class CreateDiseaseReportRequest(BaseModel):
    condition_name: str = Field(..., min_length=1, max_length=255)
    severity: DiseaseReportSeverity
    is_ai_sourced: bool = False
    ai_confidence: float | None = Field(None, ge=0, le=1)
    photo_url: str | None = Field(None, max_length=1000)


class DismissDiseaseReportRequest(BaseModel):
    dismissed_reason: str = Field(..., min_length=1)


class DiseaseReportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plant_id: uuid.UUID
    condition_name: str
    status: DiseaseReportStatus
    severity: DiseaseReportSeverity
    is_ai_sourced: bool
    ai_confidence: float | None
    photo_url: str | None
    confirmed_by_user_id: uuid.UUID | None
    confirmed_at: datetime | None
    dismissed_reason: str | None
    resolved_at: datetime | None
    created_at: datetime


class ApplyTreatmentRequest(BaseModel):
    description: str = Field(..., min_length=1)
    outcome: TreatmentOutcome


class TreatmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    disease_report_id: uuid.UUID
    description: str
    outcome: TreatmentOutcome
    applied_by_user_id: uuid.UUID
    applied_at: datetime


# ==============================================================================
# Timeline
# ==============================================================================


class PlantTimelineEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    event_type: str
    occurred_at: datetime
    summary: str
    source_id: uuid.UUID
    actor_user_id: uuid.UUID | None
